---
title: "Analyze live-dead-dying samples"
author: "Kieran R Campbell"
output:
  html_document: 
    toc: true
    toc_float: true
  html_notebook: default
params:
  cellranger_version: v3
  output_rds: 'output.rds'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      cache = FALSE)

suppressPackageStartupMessages({
  library(scater)
  library(DropletUtils)
  library(SingleCellExperiment)
  library(biomaRt)
  library(dplyr)
  library(limma)
  library(edgeR)
  library(tidyverse)
  library(ggrepel)
  library(org.Hs.eg.db)
  library(glue)
  library(here)
})
```


Read in the data:

```{r}
sce_files <- dir(glue(here("data/scesets/{params$cellranger_version}/")), full.names = TRUE, pattern = "SA928")

sce_files <- sce_files[grepl("raw", sce_files)]

sces <- lapply(sce_files, readRDS)

sces <- lapply(sces, function(sce) {
  rowData(sce)[,-c(1:2)] <- NULL
  colData(sce)$GSC_BRC <- NULL
  sce
})

sce <- do.call("cbind", sces)
```

```{r}
print(sce)
```

Print number of cells in each condition:

```{r}
print(table(sce$cell_status))
```




Add relevant info:

```{r}

rowData(sce)$ensembl_gene_id <- rownames(sce)

sce <- getBMFeatureAnnos(sce, filters = "ensembl_gene_id",
attributes = c("ensembl_gene_id", "hgnc_symbol", "entrezgene",
"start_position", "end_position", "chromosome_name"),
dataset = "hsapiens_gene_ensembl")


# Get Mitochondrial genes for QC:
mt_genes <- which(rowData(sce)$chromosome_name == "MT")
ribo_genes <- grepl("^RP[LS]", rowData(sce)$Symbol)
feature_ctrls <- list(mito = rownames(sce)[mt_genes],
                      ribo = rownames(sce)[ribo_genes])

sce <- normalize(sce)

# Calculate QC metrics
sce <- calculateQCMetrics(sce, feature_controls = feature_ctrls)
```


```{r}
plotColData(sce, x = "total_counts", y = "total_features_by_counts", colour = "cell_status")
plotColData(sce, x = "total_features_by_counts", y = "pct_counts_mito", colour = "cell_status")
```

Turn this into a proper figure for the paper:

```{r}
col_data <- as_tibble(as.data.frame(colData(sce)))

col_data <- dplyr::mutate(col_data,
                   cell_status = stringr::str_to_title(cell_status))


col_data %>% 
  ggplot(aes(x = total_features_by_counts, y = pct_counts_mito, fill = cell_status)) +
  geom_point(data = dplyr::select(col_data, -cell_status), colour = 'grey80', fill = 'black') +
  geom_point(alpha = 0.8, size = 2, shape = 21, colour = 'grey30') +
  cowplot::theme_cowplot() +
  facet_wrap(~ cell_status) +
  theme(legend.position = "none",
        strip.background = element_rect(fill = 'white')) +
  labs(x = "# genes detected", y = "% counts mitochondrial") +
  scale_fill_brewer(palette = "Accent") +
  geom_rug()
```

```{r, fig.width = 4, fig.height = 4}
cols <- c(
  "Live"="#7bccc4",
  "Dying"="#43a2ca",
  "Dead"="#0868ac"
)

col_data %>% 
  ggplot(aes(x = cell_status, y = pct_counts_mito, fill = cell_status)) +
  geom_boxplot(outlier.size = 1) +
  cowplot::theme_cowplot() +
  theme(legend.position = "none",
        strip.background = element_rect(fill = 'white')) +
  labs(x = "Cell status", y = "% counts mitochondrial") +
  scale_fill_manual(values = cols)
```

```{r}
mito_boxplot <- last_plot()
```


Run PCA first:

```{r}
sce <- runPCA(sce, ncomponents = 3)
```

PCA plot for paper

```{r}
df_pca <- as_tibble(reducedDims(sce)[['PCA']][,1:2]) %>% 
  mutate(cell_status = stringr::str_to_title(sce$cell_status))

ggplot(df_pca, aes(x = PC1, y = PC2)) +
  cowplot::theme_cowplot(font_size = 11) +
  geom_point(data = dplyr::select(df_pca, -cell_status), colour = 'grey80') +
  geom_point(aes(colour = cell_status), alpha = 0.8) +
  facet_wrap(~ cell_status) +
  scale_colour_manual(values = cols, name = "Cell status") +
  theme(legend.position = "top", 
        strip.background = element_rect(fill = 'white'),
        strip.text = element_text(face = 'bold'))
```

```{r}
pca_plot <- last_plot()
```



# Differential expression

```{r}
col_data <- dplyr::mutate(col_data,
       ct = case_when(
         pct_counts_mito > 10 & cell_status == "Dead" ~ "dead_high",
         pct_counts_mito <= 10 & cell_status == "Dead" ~ "dead_low",
         TRUE ~ "live_dying"
       ))
```

Filter genes to have > 10 total counts:

```{r}
qplot(rowData(sce)$total_counts) +
  scale_x_log10() +
  geom_vline(xintercept = 10, colour = 'red')
```


```{r}
counts_per_gene <- rowSums(as.matrix(counts(sce)))
for_de <- counts_per_gene > 10

sce_de <- sce[for_de,]


# Remove all mito genes:

sce_de <- sce_de[!grepl("^MT-", rowData(sce_de)$Symbol),]

# Remove all mito pseudogenes

sce_de <- sce_de[!grepl("^MT", rowData(sce_de)$Symbol),]

# Remove all ribo genes

sce_de <- sce_de[!grepl("^RP[L|S]", rowData(sce_de)$Symbol),]
```

Subsample while playing with figures:

```{r}
# sce_dd <- sce_de[, sce_de$cell_status %in% c("dead", "dying")]
# sce_live <- sce_de[, sce_de$cell_status == "live"]
# sce_live <- sce_live[, sample(ncol(sce_live), ncol(sce_dd))]
# 
# sce_de <- cbind(sce_dd, sce_live)
```



```{r}
col_data <- col_data[match(sce_de$Barcode, col_data$Barcode),]

count_mat_filtered <- as.matrix(counts(sce_de))

dge <- DGEList(count_mat_filtered) # , group = factor(ids))
dge <- calcNormFactors(dge)

col_data$cdr <- scale(colMeans(count_mat_filtered > 0))[,1]
design <- model.matrix(~ cdr + ct, data = col_data)
  
dge <- estimateDisp(dge, design = design)
```

Fit the model:

```{r}
fit <- glmQLFit(dge, design = design)
```


Significance testing for all contrasts

```{r}
qlf_dead_low_vs_dead_high <- glmQLFTest(fit, coef = 3)
qlf_live_dying_vs_dead_high <- glmQLFTest(fit, coef = 4)
qlf_live_dying_vs_dead_low <- glmQLFTest(fit, contrast = c(0, 0, -1, 1))

tt1 <- topTags(qlf_dead_low_vs_dead_high, n = Inf)
tt2 <- topTags(qlf_live_dying_vs_dead_high, n = Inf)
tt3 <- topTags(qlf_live_dying_vs_dead_low, n = Inf)
```


```{r}

tidy_tt <- function(tt, str) {
  tt <- as.data.frame(tt) %>% 
  rownames_to_column("ensembl_gene_id") %>% 
  as_tibble() %>% 
    mutate(comparison = str)

  tt <- mutate(tt, gene_symbol = mapIds(org.Hs.eg.db,
                                 keys=tt$ensembl_gene_id,
                                 column="SYMBOL",
                                 keytype="ENSEMBL",
                                 multiVals="first")
  )
  tt
}
```



```{r}
tt_all <- bind_rows(
  tidy_tt(tt1, "Dead low % mito vs dead high % mito"),
  tidy_tt(tt2, "Live & dying vs dead high % mito"),
  tidy_tt(tt3, "Live & dying vs dead low % mito")
)
```



```{r}
tt_annot <- group_by(tt_all, comparison) %>% 
  top_n(30, (abs(logFC))) %>% 
  ungroup()
```

```{r}
ggplot(tt_all, aes(x = logFC, y = -log10(FDR))) +
  geom_point() +
  geom_text_repel(data = tt_annot, aes(label = gene_symbol)) +
  facet_wrap(~ comparison)
```






# Gene set enrichment



```{r}
load(here("data/genesets/human_c5_v5p2.rdata"))
go_gs <- Hs.c5

entrezgene_ensembl_map <- as.list(org.Hs.egENSEMBL)

map_ids <- function(entrezgenes) {
  x <- unlist(entrezgene_ensembl_map[entrezgenes])
  names(x) <- NULL
  x
}
```

```{r}
go_gs_ensembl <- lapply(go_gs, map_ids)
names(go_gs_ensembl) <- sub("GO_", "", names(go_gs_ensembl))

idx <- ids2indices(go_gs_ensembl, id = rownames(fit))


cam1 <- camera(dge, idx, design, contrast = 3, trend.var = TRUE)
cam2 <- camera(dge, idx, design, contrast = 4, trend.var = TRUE)
cam3 <- camera(dge, idx, design, contrast = c(0, 0, -1, 1), trend.var = TRUE)
```

Add in data and plot:

```{r}
capts <- c("Dead low % mito vs dead high % mito",
          "Live & dying vs dead high % mito",
          "Live & dying vs dead low % mito")

f <- function(x, s) {
  rownames_to_column(x, "pathway") %>% 
    mutate(comparison = s)
}

df_camera <- bind_rows(
  f(cam1, capts[1]),
  f(cam2, capts[2]),
  f(cam3, capts[3])
) %>% 
  as_tibble()
```

Textual annotation

```{r}
df_txt <- group_by(df_camera, factor(comparison)) %>% 
  top_n(10, dplyr::desc(FDR))
```


And plot:

```{r}
ggplot(df_camera, aes(x = Direction, y= -log10(FDR))) +
  geom_point() +
  facet_wrap(~ comparison, scales = "free_y") +
  geom_label_repel(data = df_txt, aes(label = pathway), size = 3)
```

```{r}
df_txt <- ungroup(df_txt)

df_txt <- mutate(
  df_txt,
  pathway = forcats::fct_reorder(factor(pathway), FDR)
)

ggplot(df_txt, aes(x = pathway, y = -log10(FDR))) +
  geom_point(aes(colour = Direction), size = 2) +
  facet_wrap(~ comparison, scales = "free_x") +
  coord_flip() +
  theme_bw(base_size = 10) +
  scale_colour_manual(values = c("Down"=scales::muted('green'), "Up"=scales::muted('orange'))) +
  theme(legend.position = "top",
        strip.background = element_rect(fill = "white")) +
  labs(x = "Pathway")


```

```{r}
enrich_plot <- last_plot()
```


Sanity check:

```{r}
results <- list(
  mito_boxplot = mito_boxplot,
  pca_plot = pca_plot,
  enrich_plot = enrich_plot,
  edger_results = tt_all,
  design = design, 
  col_data = col_data
)

saveRDS(results, params$output_rds)
```


