---
title: "SA854 cell type discovery"
author: "Kieran R Campbell"
output:
  html_document: 
    toc: true
    toc_float: true
  html_notebook: default
params:
  comparison: 1
  design_file: design.csv
  output_rds: output.rds
  pathway_output_csv: output.rds
---

This document performs differential expression for comparson `params$comparison` for temperatures 6 and 37 only


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      cache = FALSE)

suppressPackageStartupMessages({
  library(scater)
  library(SingleCellExperiment)
  library(tidyverse)
  library(glue)
  library(edgeR)
  library(limma)
  library(ggrepel)
  library(org.Hs.eg.db)
  library(cowplot)
  library(mclust)
  library(scran)
  library(readxl)
})

set.seed(123L)
```

# QC the sces together

```{r}
sa854_files <- dir("../../data/scesets/", pattern = "SA854_00[1-4]_sceset_raw", full.names = TRUE)

sces <- lapply(sa854_files, readRDS)
```

Let's plot the pct mitocondrial vs total counts:

```{r}
mito_count_plots <- lapply(sces, function(sce) {
  plotColData(sce, x = "total_features", y = "pct_counts_mito") +
    labs(subtitle = paste0(sce$id[1], " at ", sce$digestion_temperature[1], "C"))
})

plot_grid(plotlist = mito_count_plots)
```

```{r}
qc_sce <- function(sce) {
  sce[, sce$total_features > 2000 & sce$pct_counts_mito < 50]
}

sces_qc <- sapply(sces, qc_sce)
```

And some PCA plots:

```{r}
sces_qc <- lapply(sces_qc, runPCA, ncomponents = 4)

pca_plots_mito <- lapply(sces_qc, plotPCA, colour_by = "pct_counts_mito")
pca_plots_features <- lapply(sces_qc, plotPCA, colour_by = "total_features")

plot_grid(plotlist = pca_plots_mito)
plot_grid(plotlist = pca_plots_features)
```

# Cluster

Cluster the SCEs in PC space and annotate with high / low mito counts

```{r}
cluster_sce <- function(sce) {
  pca <- reducedDims(sce)[['PCA']][,1:2]
  mc <- Mclust(pca, G = 2)
  mitos <- sapply(c(1,2), function(i) mean(sce$pct_counts_mito[mc$classification == i]))
  sce$cluster <- "mito-low"
  sce$cluster[mc$classification == which.max(mitos)] <- "mito-high"
  sce
}

sces_qc <- lapply(sces_qc, cluster_sce)
```

```{r}
cluster_plots <- lapply(sces_qc, plotPCA, colour_by = "cluster")
plot_grid(plotlist = cluster_plots)
```

Let's recompute the size factors:

```{r}
normalize_sce <- function(sce) {
  sce <- computeSumFactors(sce, clusters = factor(sce$cluster))
  sce
}
```


# Differential expression

```{r}
threshold_sce <- function(sce, min_counts = 50) {
  sce <- calculateQCMetrics(sce)
  sce[rowData(sce)$total_counts > min_counts, ]
}

get_vooms <- function(sce, cluster_str = "mito-high") {

  dge <- DGEList(counts(sce))
  dge <- calcNormFactors(dge, logratioTrim = 0.7, sumTrim = 0.3)
  # dge$samples$norm.factors <- sizeFactors(sce)
  
  design <- model.matrix(~cluster == cluster_str, colData(sce)) 
  
  v <- voom(dge, design, plot = TRUE)
  
  fit <- lmFit(v, design)
  fit <- eBayes(fit)
  list(fit = fit, v = v, design = design)
}
  
voom_to_df <- function(fit, sce)  {
  
  qvals <- p.adjust(fit$p.value[,2], method = 'BH')
  
  df_limma <- data_frame(id = sce$id[1], 
                         log2foldchange = fit$coefficients[,2], 
                         pval = fit$p.value[,2],
                         qval = qvals,
                         ensembl_id = rownames(sce),
                         entrezgene = rowData(sce)$entrezgene,
                         gene_symbol = rowData(sce)$Symbol,
                         chromosome = rowData(sce)$chromosome_name,
                         is_significant = qval < 0.05) 
  df_limma
}
```

```{r}
sces_qc <- lapply(sces_qc, normalize_sce)
sces_de <- lapply(sces_qc, threshold_sce)
voom_fits_and_v <- lapply(sces_de, get_vooms)
fits <- lapply(voom_fits_and_v, `[[`, 'fit')
vs <- lapply(voom_fits_and_v, `[[`, 'v')
designs <- lapply(voom_fits_and_v, `[[`, 'design')
df_des <- Map(voom_to_df, fits, sces_de)
```

Okay, so far so good

```{r}
lapply(df_des, dim)
```

```{r}
df_de <- bind_rows(df_des)

df_de$gene_type <- "other"
df_de$gene_type[df_de$chromosome == "MT"] <- "mitochondrial"
df_de$gene_type[grepl("^RP[L|S]", df_de$gene_symbol)] <- "ribosomal"
df_de$gene_type[grepl("^HLA", df_de$gene_symbol)] <- "HLA"
df_de$gene_type[grepl("^B2M", df_de$gene_symbol)] <- "HLA"
```

```{r}
cols <- c("ribosomal"="darkred", "mitochondrial"="darkgreen","other"="grey80", "HLA"="purple")

ggplot(df_de, aes(x = log2foldchange, y = -log10(qval), colour = gene_type)) + 
  geom_point(alpha = 0.8) +
  facet_wrap(~ id) +
  scale_colour_manual(values = cols) +
  labs(x = "log2foldchange mito-high vs low")
```


Camera analysis

```{r}
camera_analysis <- function(fit, v, design, df_limma, sce_de) {
  idx <- ids2indices(hallmark_genesets_ensembl, id = rownames(fit))
  cam <- camera(v, idx, design, trend.var = TRUE)

  
  pathways <- names(hallmark_genesets_ensembl)

  cam <- rownames_to_column(cam, "pathway")
  
  
  df_lfc <- lapply(pathways, function(pathway) {
    df_des[[1]][idx[[pathway]], ] %>%
    summarise(mean_log2foldchange = mean(log2foldchange), pathway = pathway)
  }) %>%
    bind_rows()
  
  df_gs <- inner_join(cam, df_lfc) %>%
    dplyr::mutate(significant = FDR < 0.05,
                  digestion_temperature = sce_de$digestion_temperature[1]) %>%
    as_tibble()
  
  df_gs
}
```


```{r}
load("../../data/genesets/human_H_v5p2.rdata")
hallmark_genesets <- Hs.H

entrez_to_ensembl <- function(entrezids) {
  entrezids <- intersect(entrezids, mappedkeys(org.Hs.egENSEMBL))
  unlist(as.list(org.Hs.egENSEMBL[entrezids]))
}

hallmark_genesets_ensembl <- lapply(hallmark_genesets, entrez_to_ensembl)
names(hallmark_genesets_ensembl) <- sub("HALLMARK_", "", names(hallmark_genesets_ensembl))

df_cams <- Map(camera_analysis, fits, vs, designs, df_des, sces_de)
```

```{r}

df_cam <- bind_rows(df_cams)

theme_set(theme_cowplot(font_size = 11))

df_cam$digestion_temperature <- paste0(df_cam$digestion_temperature, "C")

pos <- position_jitter(width = 0.05, seed = 123)
ggplot(df_cam, aes(x = Direction, y = -log10(FDR), colour = Direction)) +
  geom_point() +
  geom_label_repel(data = filter(df_cam, FDR < 0.01), aes(label = pathway), size = 2) +
  facet_wrap(~ digestion_temperature, scales = "free_y") +
  scale_colour_manual(values =c("Up"="darkgreen", "Down"="darkred")) +
  theme(legend.position = "none",
        strip.background = element_rect(fill = "grey90")) +
  labs(x = "Regulation in mitochondrial-'hot' cell type")

ggsave("../../figs/sa854_mito_vs_rest_enrichment.png", width = 7, height = 5)
```

Let's just graph the oxidative phosphorylation at 6C:

```{r}
sce <- sces_de[[1]]
# op_genes <- hallmark_genesets_ensembl[["OXIDATIVE_PHOSPHORYLATION"]]
op_genes <- hallmark_genesets_ensembl[["MYC_TARGETS_V1"]]
op_in_sce <- intersect(op_genes, rownames(sce))
sce_op <- sce[op_in_sce,]
```


```{r}
filter(df_des[[2]], ensembl_id %in% op_in_sce) %>% 
  ggplot(aes(x = log2foldchange)) +
  geom_histogram()
```

# Joint analysis

```{r}
rowdata_to_strip <- names(rowData(sces_de[[1]]))[13:20]

strip_rowdata <- function(sce) {
  rowData(sce)[rowdata_to_strip] <- NULL  
  sce
}

common_genes <- Reduce(intersect, lapply(sces_de, rownames))

sces_de_cg <- lapply(sces_de, function(sce) {
  sce[common_genes,]
})

sces_de_cg <- lapply(sces_de_cg, strip_rowdata)

sce_all <- do.call("cbind", sces_de_cg)
```

Time for some DE!

```{r}
dge <- DGEList(counts(sce_all))
dge <- calcNormFactors(dge, logratioTrim = 0.7)

sce_all$numeric_temp <- as.vector(scale(as.numeric(sce_all$digestion_temperature)))
sce_all$cluster <- factor(sce_all$cluster, levels = c("mito-low", "mito-high"))
  
design <- model.matrix(~ cluster * numeric_temp, colData(sce_all)) 
  
v <- voom(dge, design, plot = TRUE)
  
fit <- lmFit(v, design)
fit <- eBayes(fit)

qvals <- apply(fit$p.value, 2, p.adjust, method = "BH")

df_limma <- data_frame(log2foldchange_cluster = fit$coefficients[,2], 
                       log2foldchange_temp = fit$coefficients[,3],
                       log2foldchange_cluster_temp = fit$coefficients[,4],
                       qval_cluster = qvals[,2],
                       qval_temp = qvals[,3],
                       qvals_cluster_temp = qvals[,4],
                       ensembl_id = rownames(sce_all),
                       entrezgene = rowData(sce_all)$entrezgene,
                       gene_symbol = rowData(sce_all)$Symbol,
                       chromosome = rowData(sce_all)$chromosome_name)
df_limma
```


```{r}
ggplot(df_limma, aes(x = log2foldchange_cluster, y = -log10(qval_cluster))) +
  geom_point()
```

```{r}
top_temp <- top_n(df_limma, 20, abs(log2foldchange_temp))

df_limma$gene_type <- "other"
df_limma$gene_type[grepl("^HSP", df_limma$gene_symbol)] <- "Heat shock protein"

ggplot(df_limma, aes(x = log2foldchange_temp, y = -log10(qval_temp), colour = gene_type)) +
  geom_point() +
  geom_text_repel(data = top_temp, aes(label = gene_symbol), size = 3, colour = 'black') +
  scale_colour_manual(values = c("other"="grey90","Heat shock protein"="darkred"), name = "Gene type") +
  geom_point(data = filter(df_limma, gene_type == "Heat shock protein"))
```

```{r}
top_int <- top_n(df_limma, 20, abs(log2foldchange_cluster_temp))

ggplot(df_limma, aes(x = log2foldchange_cluster_temp, y = -log10(qvals_cluster_temp))) +
  geom_point() +
  geom_text_repel(data = top_int, aes(label = gene_symbol), size = 3)
```

Save df limma to file:

```{r}
dplyr::select(df_limma, ensembl_id, gene_symbol, entrezgene, chromosome, gene_type, everything()) %>% 
  write_csv("../../data/differential_expression/sa854_de_all.csv")
```

Cell surface markers

```{r}
surface_markers <- read_excel("../../data/genesets/S2_File.xlsx")
```

```{r}
surface_markers <- filter(surface_markers, `CSPA category` == "1 - high confidence") %>% 
  .$UP_entry_name

surface_markers <- gsub("_HUMAN", "", surface_markers, fixed = FALSE)
```

```{r}
dfl_sm <- filter(df_limma, gene_symbol %in% surface_markers)

dfl_sm <- filter(dfl_sm, qval_cluster < 0.01)
```

And plot

```{r}
hist(dfl_sm$log2foldchange_cluster)
```


```{r}
dfl_sm_top <- top_n(dfl_sm, 20, abs(log2foldchange_cluster)) %>% 
  arrange((log2foldchange_cluster))

dfl_sm_top$gene_symbol <- factor(dfl_sm_top$gene_symbol, levels = dfl_sm_top$gene_symbol)

ggplot(dfl_sm_top, aes(x = gene_symbol, y = log2foldchange_cluster)) +
  geom_bar(stat = "identity") +
  coord_flip()
```






Camera analysis

```{r}
idx <- ids2indices(hallmark_genesets_ensembl, id = rownames(fit))
cam_cluster <- cameraPR(fit$coefficients[,2], idx)
cam_temp <- cameraPR(fit$coefficients[,3], idx)
cam_inter <- cameraPR(fit$coefficients[,4], idx)
```

